### PR TITLE
Add `RecordSet` support for route53 controller

### PR DESCRIPTION
**Pull request by kunhwiko**
**Issue**: 
https://github.com/aws-controllers-k8s/community/issues/1545 

**Description of changes**:
Adds support for RecordSets in Route53 and corresponding E2E tests. This includes support for all record types and routing policies.

Some previous conversations for how to implement this were discussed here:
- https://kubernetes.slack.com/archives/C0402D8JJS1/p1694054604699869
- https://kubernetes.slack.com/archives/C0402D8JJS1/p1695030078194489

Changes made to following files/directories (other changes are all code-gen'ed):
- generator.yaml
- pkg/resource/record_set/hooks.go
- templates/hooks/record_set/
- test/e2e/

Generator results:
```
make build-controller SERVICE=route53
building ack-generate ... ok.
==== building route53-controller ====
Copying common custom resource definitions into route53
Building Kubernetes API objects for route53
Generating deepcopy code for route53
Generating custom resource definitions for route53
Building service controller for route53
Generating RBAC manifests for route53
Running gofmt against generated code for route53
Updating additional GitHub repository maintenance files
==== building route53-controller release artifacts ====
Building release artifacts for route53-v0.0.8
Generating common custom resource definitions
Generating custom resource definitions for route53
Generating RBAC manifests for route53
```

E2E results:
```
[gw1] [ 25%] PASSED tests/test_hosted_zone.py::TestHostedZone::test_create_delete_private
[gw0] [ 50%] PASSED tests/test_hosted_zone.py::TestHostedZone::test_create_delete_public
[gw2] [ 75%] PASSED tests/test_hosted_zone.py::TestHostedZone::test_crud_tags
[gw4] [100%] PASSED tests/test_record_set.py::TestRecordSet::test_crud_simple_record
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: generator.yaml
Comment (kunhwiko) [inline generator.yaml:26]: RecordSets are a little odd because we have to model them from `ChangeResourceRecordSets`:
- https://github.com/aws/aws-sdk-go/blob/main/models/apis/route53/2013-04-01/api-2.json#L1299

This operation takes in batches of multiple record sets at the same time and various actions (e.g. CREATE, DELETE, UPSERT) to specify what to do with these record sets. I decided to model the RecordSet resource as an individual CR instead of having a single CR representing multiple record sets to make it simpler for reconciliation. 
-

File: generator.yaml
Comment (kunhwiko) [inline generator.yaml:78]: It seems that RecordSets don't really have a single field that represents a unique identifier. Instead, multiple fields are used to guarantee the uniqueness of a record set:
- DNSName
- HostedZoneID
- RecordType
- SetIdentifier 

Here, I've set the ID as the primary key for now. This is a `ChangeID` that represents changes made when `ChangeResourceRecordSets` is invoked. This ID is used in our reconcile logic to get the current status of a particular change (i.e. have the changes been synced to all DNS servers).

File: pkg/resource/record_set/hooks.go
Comment (kunhwiko) [inline pkg/resource/record_set/hooks.go:1]: Since `ChangeResourceRecordSets` takes in a ChangeBatch (i.e. batches of record sets), this file mostly is comprised of of helper functions that are used to shape our CR specs into a ChangeBatch object.

File: pkg/resource/record_set/utils.go
Comment (kunhwiko) [inline pkg/resource/record_set/utils.go:18]: While testing simple and alias records, I've noticed ListResourceRecordSets seems to always append a `.` to the end of a DNS name even if we've supplied a CR that does not have the trailing `.` (not the case for CNAME values though). Below is an example CR

```
apiVersion: route53.services.k8s.aws/v1alpha1
kind: RecordSet
metadata:
  name: alias-record
spec:
  name: "*.some.random.example.ack.com"
  hostedZoneID: <hosted-zone-id>
  recordType: A
  aliasTarget:
    evaluateTargetHealth: false
    dnsName: vpce-something-random.com
    hostedZoneID: <vpc-endpoint-hosted-zone-id>
```

In the above case for `spec.name`, the result that we get from `ListResourceRecordSets` would be `\\052.some.random.example.ack.com.`

File: test/e2e/tests/test_record_set.py
Comment (kunhwiko) [inline test/e2e/tests/test_record_set.py:100]: Test results:
```
[gw1] [ 25%] PASSED tests/test_hosted_zone.py::TestHostedZone::test_create_delete_private
[gw0] [ 50%] PASSED tests/test_hosted_zone.py::TestHostedZone::test_create_delete_public
[gw2] [ 75%] PASSED tests/test_hosted_zone.py::TestHostedZone::test_crud_tags
[gw4] [100%] PASSED tests/test_record_set.py::TestRecordSet::test_crud_simple_record
```

File: test/e2e/tests/test_hosted_zone.py
Comment (kunhwiko) [inline test/e2e/tests/test_hosted_zone.py:14]: Changes in this file are essentially the same, just refactoring some parts because there was a lot of common code with `test_record_set.py`

File: apis/v1alpha1/generator.yaml
Comment (a-hilaly) [inline apis/v1alpha1/generator.yaml:140]: Now that we support [drift remediation configuration via flags/helm values](https://aws-controllers-k8s.github.io/community/docs/user-docs/drift-recovery/), we can drop this.

File: apis/v1alpha1/generator.yaml
Comment (a-hilaly) [inline apis/v1alpha1/generator.yaml:110]: Can we make move it the status? or is this something the user can input. Could we consider randomizing it value?

File: apis/v1alpha1/types.go
Comment (a-hilaly) [inline apis/v1alpha1/types.go:71]: If i understand well this field is not really be provided by the user. We can deduce the necessary changes we need to make from CR deltas, and generate the right `Changes` object.

File: pkg/resource/record_set/utils.go
Comment (a-hilaly) [inline pkg/resource/record_set/utils.go:18]: Great one!

Comment (a-hilaly) [reply pkg/resource/record_set/utils.go:18]: Can we put this function in `hooks.go` - not against `utils.go` naming but it just a standard in almost all the ACK controllers.

File: pkg/resource/record_set/sdk.go
Comment (a-hilaly) [inline pkg/resource/record_set/sdk.go:253]: How about making this part of the `sdkFind`, anyways..

```suggestion
		err = rm.syncStatus(ctx, ko)
		if err != nil {
			return nil, err
		}
		// Also force a requeue if the Change is still pending
		if sync.Status.Status != *"PENDING" {
		    ackcondition.SetSynced(&resource{ko}, corev1.ConditionFalse, nil, nil)
		}
```

File: pkg/resource/record_set/hooks.go
Comment (a-hilaly) [inline pkg/resource/record_set/hooks.go:183]: 👍

File: apis/v1alpha1/record_set.go
Comment (a-hilaly) [inline apis/v1alpha1/record_set.go:399]: This should be generated as `ttl` ... maybe the regex is messed up here: https://github.com/aws-controllers-k8s/pkg/blob/main/names/names.go#L167

Comment (a-hilaly) [reply apis/v1alpha1/record_set.go:399]: Bumping the code-generator version could resolve this

File: test/e2e/tests/test_record_set.py
Comment (a-hilaly) [reply test/e2e/tests/test_record_set.py:100]: Good testing @kunhwiko !

File: test/e2e/resources/record_set_simple.yaml
Comment (a-hilaly) [inline test/e2e/resources/record_set_simple.yaml:11]: i believe this should be `ttl`

File: test/e2e/tests/test_record_set.py
Comment (a-hilaly) [inline test/e2e/tests/test_record_set.py:120]: nit: This is patching the `resourceRecords` not the TTL

**Review by a-hilaly — COMMENTED** [review]
@kunhwiko Fantastic job on this! Really digging the neat code.

I have a few suggestions, comments, and questions below. I'm also interested in cloning your branch locally so I can experiment with the Route 53 API and do some local tests.

I will very likely review again the tests/hooks.go before EOW.

File: apis/v1alpha1/generator.yaml
Comment (kunhwiko) [reply apis/v1alpha1/generator.yaml]: @a-hilaly Didn't quite make the suggested change as the intent here was to mention that the ID refers to the ChangeID. Could you check to see if the changed documentation looks ok?

File: apis/v1alpha1/generator.yaml
Comment (kunhwiko) [reply apis/v1alpha1/generator.yaml:51]: So, I might have misunderstood the statement here. If we ignore this field in the generator, problem is the path to all the other fields will be ignored as well (e.g. `ChangeBatch.Changes.ResourceRecordSet.ResourceRecords`). Unfortunately, `ResourceRecordSet` is embedded deep within this ChangeBatch shape and I couldn't find an alternative

File: test/e2e/resources/record_set_simple.yaml
Comment (a-hilaly) [reply test/e2e/resources/record_set_simple.yaml:11]: i will send a patch to `github.com/aws-controllers/pkg` and `code-generator`. Need to hold until we fix this :)
/hold

File: apis/v1alpha1/record_set.go
Comment (a-hilaly) [reply apis/v1alpha1/record_set.go:398]: I'll work on fixing this... i'll ping you once we have a fix merged in code-generator

**Comment by PT-GD** [comment]
> @kunhwiko I opened [aws-controllers-k8s/pkg#23](https://github.com/aws-controllers-k8s/pkg/pull/23) - awaiting for another ACK maintainer to ship it and we'll have in code-gen soon

Looks like it got merged!

**Comment by kunhwiko** [comment]
@PT-GD Oh yeah, you're right - I'll look to regenerate code for this PR shortly

**Comment by PT-GD** [comment]
> @PT-GD Oh yeah, you're right - I'll look to regenerate code for this PR shortly :)

Any update?

**Comment by kunhwiko** [comment]
@PT-GD Hey there, I didn't get the chance last week since I was traveling - will get to this hopefully by Monday

**Comment by a-hilaly** [comment]
Expecting to merge this week. Just want to figure out if there is a way we can omit the RecordChange type from the Spec...

**Review by a-hilaly — APPROVED** [review]
Thank you very much for your work on this @kunhwiko !
/approve
/lgtm

I still think we need to find a way to remove the ChangeBatch field but we'll have to tweak the code generator a little for that. We can address this in a different patch.

**Pull request merged**
